### PR TITLE
MULE-18976: Implement Feature Flagging 

### PR DESCRIPTION
**Pull request by szaffarano**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistryTestCase.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.tck.size.SmallTest;
+
+@SmallTest
+public class FeatureFlaggingRegistryTestCase {
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistryTestCase.java:24]: allure

File: core-tests/src/test/java/org/mule/runtime/core/internal/config/DefaultFeatureFlaggingServiceTestCase.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+@RunWith(Parameterized.class)
+public class DefaultFeatureFlaggingServiceTestCase {
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/config/DefaultFeatureFlaggingServiceTestCase.java:31]: allure

File: core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ *    private FeatureFlaggingService featureFlaggingService;
+ *    // ....
+ *    
+ *    if (featureFlaggingService.isEnabled("some feature")) {
+ *        // ...
+ *    }
+ * </code>
+ * 
+ * @see FeatureFlaggingRegistry
+ * @since 4.4.0
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java:41]: will this be cherrypicked?

File: core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingService.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * application will have its own set of flags independently of the rest of the applications deployed in a given runtime.
+ * 
+ * @see FeatureFlaggingRegistry
+ * @since 4.4.0
+ */
+public interface FeatureFlaggingService {
+
+  String FEATURE_FLAGGING_SERVICE_KEY = "core.featureFlaggingService";
+
+  boolean isEnabled(String feature);
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingService.java:22]: add javadoc

File: core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ */
+public class FeatureFlaggingRegistry {
+
+  private final Map<String, Predicate<MuleContext>> configurations = new ConcurrentHashMap<>();
+
+  private static final FeatureFlaggingRegistry INSTANCE = new FeatureFlaggingRegistry();
+
+  public static final FeatureFlaggingRegistry getInstance() {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java:49]: javadoc

File: core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    return INSTANCE;
+  }
+
+  private FeatureFlaggingRegistry() {}
+
+  public void registerFeature(String feature, Predicate<MuleContext> condition) {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java:55]: javadoc

File: core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      throw new MuleRuntimeException(createStaticMessage("Error registering %s: condition must not be null", feature));
+    }
+
+    Predicate<MuleContext> added = configurations.putIfAbsent(feature, condition);
+    if (added != null) {
+      throw new MuleRuntimeException(createStaticMessage("Feature %s already registered", feature));
+    }
+  }
+
+  public Map<String, Predicate<MuleContext>> getFeatureConfigurations() {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java:70]: javadoc

File: core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    return unmodifiableMap(configurations);
+  }
+
+  protected void clearFeatureConfigurations() {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java:74]: javadoc. Is this just for test?

File: core/src/main/java/org/mule/runtime/core/api/config/MuleConfiguration.java [hunk 9]
```diff
@@ -7,15 +7,18 @@
...
 import org.mule.runtime.core.api.construct.Flow;
 import org.mule.runtime.core.api.processor.strategy.ProcessingStrategy;
 import org.mule.runtime.core.api.processor.strategy.ProcessingStrategyFactory;
 
+import java.util.Optional;
+
 /**
- * Configuration info. which can be set when creating the MuleContext but becomes immutable after startup.
- * TODO MULE-13121 Cleanup MuleConfiguration removing redundant config in Mule 4
+ * Configuration info. which can be set when creating the MuleContext but becomes immutable after startup. TODO MULE-13121 Cleanup
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/config/MuleConfiguration.java:20]: keep these lines separated with `<p>`

File: tests/test-extensions/petstore-extension/src/main/java/org/mule/test/petstore/extension/PetStoreOperations.java [hunk 10]
```diff
@@ -70,6 +89,14 @@ public String echoWithSignature(String message) {
     return message + " echoed by Petstore";
   }
 
+  @MediaType(TEXT_PLAIN)
+  public String featureFlaggedEcho(String message) {
+    if (ffService.isEnabled(LEGACY_BEHAVIOUR)) {
+      return String.format("%s [old way]", message);
```
Comment (elrodro83) [inline tests/test-extensions/petstore-extension/src/main/java/org/mule/test/petstore/extension/PetStoreOperations.java:95]: static import

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java (hunk 3 above)
Comment (szaffarano) [reply core/src/main/java/org/mule/runtime/core/api/config/FeatureFlaggingRegistry.java:41]: Yes, I think we should backport it to 4.3.0 and perhaps 4.2.x/4.2.2. Do you agree?

**Review by elrodro83 — APPROVED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**